### ISSUE TITLE
Call for testing open source setup - round 2

### ISSUE THREAD

**Issue by pal1000** [pal1000/Realtek-UAD-generic#29]
Releases 6.0.8873.1 through 6.0.8876.1 include a beta version of open source setup - `setup-beta.cmd` in addition to stable release. The new setup changes how admin rights are requested. It uses Powershell instead of vbscript to simplify elevation block; that convoluted and hard to read part of code that continues to fail in edge case environments with last installment of it being #28. @purifiedh2o did a pretty good job in #18, but if it would have been flawless we wouldn't have had #28.

Ping users who contributed code, experienced hiccups with Realtek Audio Console, reported valid bugs or provided good suggestions against setup or who had confirmed conflicting drivers in Windows driver store:
- [x] @HumbertoFreitas (reason: reported a crash in elevation block - #28, that is confirmed to be fixed in beta);
- [x] @gtumanyan (reason: conflicting drivers in Windows driver store - #26);
- [ ] @AlyoshaVasilieva (reason: conflicting drivers in Windows driver store - #26);
- [x] @akks1981 (reason: bug report - #25);
- [x] @DooGie666 (reasons: enhancement request - #7, bug report - #24);
- [ ] @lllopo (reason: bug report - #21);
- [ ] @julianxhokaxhiu (reasons: hiccup with Realtek Audio Console due to conflict in Windows driver store - #2, bug report - #17, bug resolution collaborator - #20);
- [ ] @purifiedh2o (reason: code contribution - #18);
- [ ] @suppai (reason: hiccup with Realtek Audio Console due to unknown reasons - #15);
- [ ] @madakab (reason: hiccups with Realtek Audio Console due to missing, conflicting or duplicate drivers in Windows driver store - #6, #12);
- [x] @Velken (reason:  hiccup with Realtek Audio Console and recurrent conflict in Windows driver store - #5).

Based on on discussion so far I created the following TODO list:
- [x] Print a warning about potential spontaneous restarts before doing anything - DONE, shipped in 6.0.8876.1 release (requested by @Velken);
- [x] Stop Windows Audio service when performing force update - DONE, shipped in 6.0.8876.1 release (should help @Velken and @zahbod);
- [x] Re-implement waits using CHOICE for network-less compatibility - DONE, shipped in 6.0.8876.1 release (should help @gtumanyan and @Velken);
- [x] Use powershell to request admin rights in force updater when using beta setup - DONE, shipped in 6.0.8876.1 release.

**Comment by julianxhokaxhiu** [comment]
I did install using normal `setup.cmd` and everything went fine. Didn't try the new beta setup though.

**Comment by DooGie666** [comment]
The beta installer worked fine as did the standard installer. However I always run as admin from the context menu so not double clicking.
Checked the driver store with RAPR and no conflicting drivers found.

**Comment by akks1981** [comment]
The beta installer worked 100% for me. The Standard installer works good to. No bugs!

**Comment by purifiedh2o** [comment]
Unfortunately my new PC forces using INTELAUDIO (Intel Smart Sound Technology integration) HWID instead of HDAUDIO, I can no longer test this script now.

**Comment by pal1000** [comment]
My package includes hardware IDs for INTELAUDIO in `HDXRTSST.inf` and `HDX_GenericExt_RTK.inf` so it could still work.

**Comment by Velken** [comment]
Installer works fine, just had some files not found issues, felt it went faster than the previous one.
![image](https://user-images.githubusercontent.com/8467292/71900695-0ed7a100-313d-11ea-9793-7203257f35e8.png)

#5 still happening tho (and Windows Update is completely blocked from updating any drivers). It prevents access to the Audio Console, but the audio prompt for "whad did you plug?" headset/headphone/microphone still appears, so it doesn't prevent usage, since other settings can be made via Control Panel > Sound.

**Comment by pal1000** [comment]
@Velken, I added ability to remove OEM components in my previous setup update. Create a file named `oem.ini` containing `hdxsst4da.inf` on same directory as `setup.cmd` to hopefully avoid #5.

Do you get those errors with `setup.cmd` or you only get them with `setup-beta.cmd`?

**Comment by gtumanyan** [comment]
Got an infinite 20 seconds with beta installer

![image](https://user-images.githubusercontent.com/24494814/71916972-22602780-3190-11ea-9a5d-fb002c149736.png)
Pressed anykey
![image](https://user-images.githubusercontent.com/24494814/71917067-4e7ba880-3190-11ea-896c-f0716409f436.png)
Then everything installed just fine without orphaned drivers left.

**Comment by Velken** [comment]
@pal1000 tried the .ini with beta, it uninstalls and installs everything correctly, it worked and can open Audio Console, BUT ONLY if I skip the "UAD Generic Driver Updater", as it restarts without warning¹ and **doesn't install**. 
Tried a few times and nothing. Using your default setup also doesn't work. 

Removing the .ini makes the forced updater install correctly, but then breaks Audio Console.

That aside, great work!


P.S.: Still getting the file not found with the beta installer (with and without the .ini)


1- Shouldn't there be a pause to press any key before restarting or at least a warning in the first line? To ensure user saved anything important

**Comment by pal1000** [comment]
> Got an infinite 20 seconds with beta installer
> ![image](https://user-images.githubusercontent.com/24494814/71916972-22602780-3190-11ea-9a5d-fb002c149736.png)
> Pressed any key then everything installed just fine without orphaned drivers left.

I have no idea why it hanged [here](https://github.com/pal1000/Realtek-UAD-generic/blob/977615aad86f86d12977113a90b2e456b7906f3a/setup-beta.cmd#L124).

**Comment by pal1000** [comment]
@Velken, there are some things that seam worrisome about your system:
- errors only in beta when code gets on line [61](https://github.com/pal1000/Realtek-UAD-generic/blob/977615aad86f86d12977113a90b2e456b7906f3a/setup-beta.cmd#L61) or [63](https://github.com/pal1000/Realtek-UAD-generic/blob/977615aad86f86d12977113a90b2e456b7906f3a/setup-beta.cmd#L63) which may indicate corruption in Windows driver store but I hope I am wrong;
- `oem.ini` being always needed to eliminate `hdxsst4da.inf`` which normally should only be needed
once until Windows is reinstalled or upgraded;
- spontaneous reboot in install phase indicate something is still running even after uninstall is done.

I added your request about printing warning about potential spontaneous reboot before starting setup to a TODO list.

**Comment by Velken** [comment]
@pal1000 ok, I have another four laptops I can test, two from same manufacturer and series (Acer, series Nitro) and two completely different, I'll post the results I got once I am done.

**Comment by HumbertoFreitas** [comment]
I get this error every time I am uninstalling (I get no errors beyond this).

![](https://user-images.githubusercontent.com/54781225/71939045-1e340a00-3190-11ea-969b-044bb95e020e.png)

**Comment by pal1000** [comment]
> I get this error every time I am uninstalling (I get no errors beyond this). 
> ![](https://user-images.githubusercontent.com/54781225/71939045-1e340a00-3190-11ea-969b-044bb95e020e.png)

I wonder if beta setup gets stuck in infinite loop if lines 40-45 are removed.
Anyway, I can take a look. For starters I only need a driver list dump from [Driver Store Explorer](https://github.com/lostindark/DriverStoreExplorer/releases). Run it, do `File` - `Export Driver List`, zip then attach hhe resulted `.csv` file.

**Comment by zahbod** [comment]
beta installer worked perfectly for me except one thing although i'm a bit confused that maybe this isn't the beta installer but this is definitely a problem in installation. after beta installer did it's thing and it was force updater turn it restarted for disabling device which is fair but after reboot the device wasn't disabled i tried it a couple of times and the realtek console still showed the 6.0.8856.1 approved it so in the end i force updated the codecs manually and i wanted you to know that for some reason the force updater script doesn't work properly

**Comment by pal1000** [comment]
@zahbod, I haven't heard of anyone experiencing a reboot in force updater, but I'll look into it.

**Comment by Velken** [comment]
@pal1000 as I stated earlier, I had reboot on the forced updater. And it did not try to run again after reboot. Also, here is what I gathered from the tests on other machines:

The two non-Acer worked fine, one was hanging on the 'ping' wait 20 seconds. I changed it to a 'REM|CHOICE' wait 20 seconds and worked fine (this is an old-ish  Avell laptop that has no network drivers or connections.

All Acer had same issue with the cannot find file, I checked the folder, it was encrypted (similar to WindowsApps folder/UWP encryption)

All Acer had the exact same need to use oem.ini (all Windows 10 copies are direct from factory on those four, so I can rule out it being something on my machine), they all got the [...]"doesn't support this machine!" (when doing the forced updater) each time I restarted the computer or made any changes that temporarily disabled all drivers and then re-enabled them. Skipping forced updater made the Audio Console work on them.

All Acer had the spontaneuos reboot on forced updater with oem.ini

My guess is that is one of Acer's "special softwares" messing stuff up, since it only happened on those, but a (n=3) is not a good enough pool to make sure assumptions.

**Comment by zahbod** [comment]
> 
> 
> @zahbod, I haven't heard of anyone experiencing a reboot in force updater, but I'll look into it.
> 
> PS: Did force updater tried to run again after reboot? Because that's the expected behavior.

no it didn't tried to run again after reboot

**Comment by pal1000** [comment]
@zahbod, @Velken, I think force updater actually succeeds both disable and enable commands, but it fails the overwrite command and the enable command is the one that triggers the reboot. It's the only scenario I can think of where force updater would fail and not re-run on next reboot.

**Comment by zahbod** [comment]
@pal1000 Because i know how to manually force update i can say confidently that no it failed to disable after a couple of times i tried i said welp i guess i should do it manually and then the actual codec was locked by the service and device so it wasn't disabled to unlock the file and replace it

**Comment by pal1000** [comment]
@zahbod, I think I didn't made myself very clear. I meant it succeeds both disable and enable commands, but due to overwrite failure which is crucial for force updater to actually do something it appears like nothing happened. You would still be on 6.0.8856.1 (the WHQL base) and it would be locked and force updater won't retry running either which is exactly what you and @Velken described as being your problem.

**Comment by pal1000** [comment]
6.0.8876.1 contains the following setup improvements:
- Re-implementation of waits using CHOICE instead of ping for compatibility with systems lacking networking;
- Force updater now stops Windows Audio service before overwriting files to reduce failure likelihood;
- Added a warning about potential spontaneous restarts to both main setup and force updater;
- There is a beta setup paired by a beta force updater included that uUse powershell to request admin rights to workaround #28.

**Comment by nbarsotti** [comment]
@pal1000 , First of all I want to thank you for creating this project and making it available.  It has been immensely helpful to me.  This question is a out of scope compared to this thread, but I did not know how else to contact you.  My question is this:  Is it possible to use direct Realtek generic downloads from their FTP site (I have access to the downloads) to install the driver on a motherboard that does support UAD drivers from the manufacturer?

Thank you for your consideration, and if its possible to take this conversation out of this thread please let me know.

**Comment by pal1000** [comment]
@nbarsotti, the short answer is yes, you can, but those aren't generic packages, they are OEM specific so out of this protect scope. The quest of finding the right package begins with identification of sound card hardware ID. Your question is better suited on forums like station-drivers, tenforums or mydigitallife. You can also search your sound card hardware ID [here](https://github.com/alanfox2000/realtek-universal-audio-driver). This will give you a list of `.inf` files that must be in the package in order for it to be compatible with your sound card.

**Comment by gtumanyan** [comment]
The beta updater worked flawlessly. 
On my system installation of the new audio driver doesn't trigger automatically download UWP/HSA from Store so I had to manually download and install it from [here].(https://github.com/alanfox2000/realtek-universal-audio-driver/releases)

**Comment by pal1000** [comment]
@gtumanyan, only Realtek decides which releases update HSA/UWP. UWP is updated only when newer driver package ships an updated HSA, This last happened in 6.0.8874.1. If you were running 6.0.8874.1 and then updated to 6.0.8876.1 there is no HSAUWP update.

**Comment by pal1000** [comment]
Beta testing ends with 6.0.8878.1

@HumbertoFreitas, you seam to be the only one who didn't respond. Please open another issue with information requested [here](https://github.com/pal1000/Realtek-UAD-generic/issues/29#issuecomment-571969957) if you still need help with that Realtek service not terminating.

**Comment by julianxhokaxhiu** [comment]
Just want to double confirm that I did install the latest release with the new setup and everything worked fine. Well done, and thank you!
